controllers/events_controller.js

```diff
@@ -0,0 +1,268 @@
+const kue = require("../config/Scheduler/kue");
+const worker = require("../config/Scheduler/worker");
+
+const ObjectId = require("mongoose").Types.ObjectId;
+
+// import http status codes
+const { BAD_REQUEST, NOT_ACCEPTABLE } = require("../utility/statusCodes");
+// import constants
+const { USER_HASH_LENGTH } = require("../config/index");
+// import helper functions
+const {
+	sendError,
+	sendSuccess,
+	generateHash,
+	escapeRegex
+} = require("../utility/helpers");
+
+module.exports.getParticipants = async (req, res) => {
+	let { eventId, query, branch, year, sortBy } = req.query;
+    let filters = {};
+    
+    if(eventId){
+		filters["events.event"] = new ObjectId(eventId)
+	};
+
+	if (query) {
+		const regex = new RegExp(escapeRegex(query), "gi");
+		filters.$or = [{ name: regex }, { email: regex }, { branch: regex }];
+	}
+	if (branch) {
+		filters.branch = branch;
+	}
+	if (year) {
+		filters.year = Number(year);
+	}
+
+	let sortObj = {};
+	if (sortBy) {
+		sortBy.map(sort => {
+			if (sort === "createdAt") {
+				sortObj[sort] = "desc";
+			} else {
+				sortObj[sort] = "asc";
+			}
+		});
+	} else {
+		sortObj = {
+			createdAt: "desc",
+			branch: "asc",
+			year: "asc",
+			name: "asc"
+		};
+	}
+
+	let participants = await Participant.find(filters).sort(sortObj);
+	sendSuccess(res, participants);
+};
+
+module.exports.registerParticipant = async (req, res) => {
+	let { name, email, branch, year, phone } = req.body;
+	let participant = await Participant.findOne({
+		$or: [
+			{ email: { $regex: `^${email}$`, $options: "i" } },
+			{
+				$and: [
+					{ name: { $regex: `^${name}$`, $options: "i" } },
+					{ branch },
+					{ year },
+					{
+						$or: [
+							{ email: { $regex: `^${email}$`, $options: "i" } },
+							{ phone }
+						]
+					}
+				]
+			}
+		]
+	});
+
+	if (participant) {
+		sendError(res, "Already Registered!!", BAD_REQUEST);
+	} else {
+		let password = await generateHash(USER_HASH_LENGTH);
+		participant = new Participant({
+			name,
+			email,
+			branch,
+			year,
+			phone,
+			password,
+			events: []
+		});
+		participant = await participant.save();
+		let args = {
+			jobName: "sendLoginCreds",
+			time: Date.now(),
+			params: {
+				email,
+				password,
+				name,
+				role: "Participant"
+			}
+		};
+		kue.scheduleJob(args);
+		sendSuccess(res, participant);
+	}
+};
+
+module.exports.updateParticipant = async (req, res) => {
+	let { name, email, branch, year, phone } = req.body;
+	let updateObj = {
+		name,
+		email,
```

Comment:
Should we allow updating of email and phone?
What if someone updates their email to something that is already being used?

Reply:
no

Reply:
Another scenario: What if a user enters a wrong email/phone by mistake and wants to update it later?

Reply:
we will update directly in db

Reply:
Let's think of some way to verify the email when someone updates it.